### ISSUE TITLE
Bias adjustment - added features

### ISSUE THREAD

**Issue by kellijohnson** [ss3sim/ss3sim#124]
It would be great if we could implement a way to check if a specified number of runs inverted before calculating the bias adjustment. Users could specify a number of runs that they need to converge (e.g. bias.need = 10) and the maximum runs they are willing to perform before they deem the scenario a failure (e.g. bias.max = 100). Right now I had to implement my own wrapper function that does this for the "M" paper. I run a set number of bias adjustment runs (i.e. nbias = 10) and if they I do not get 8 that were successful I delete the scenario folder and run incrementally more runs until I either reach my max or 8 invert. This wastes a large amount of computing time because you must either set your number of initial runs to a very large number or delete a lot of runs when things fail to produce your desired number.

Let me know if I am not being clear on what I would like. Furthermore, it would be great if these could be additional arguments so that users could still specify the number of runs as they are currently doing or specify a number that they want to converge.

-KFJ

**Comment by kellijohnson** [comment]
I am looking for ideas on how to implement increasing the number of bias adjustment runs while in the loop of ss3sim_base if the original number of bias adjustment runs did not meet a minimum convergence criteria. I wrote some code that will check 
```r
check_bias <- function(dir = getwd(), conv_crit = 0.2, bias_max = 20) {
  filepath <- dir(dir, pattern = "Adjust", recursive = TRUE)
  convergence <- read.table(filepath, sep = " ")
  perctnohess <- sum(is.na(convergence$V2)) / length(convergence$V2)
  if (perctnohess < conv_crit) {
    return(TRUE)
  } else {
    return(ifelse(length(convergence$V2) < bias_max, FALSE, TRUE))
    }
}
```
but the loop is ```for (i in iterations)``` where ```iterations``` includes both bias and normal runs. So if the convergence criteria returns a false I have no way to update ```iterations``` while in the loop. The only thing I can think of is to change the for loop to a while loop but I wanted to get other people's opinions.

**Comment by kellijohnson** [comment]
Just thought of an alternative to ```while```. You could make the number of bias adjustment iterations equal to ```bias_max``` check convergence at ```bias_nsim``` and keep running through the iterations if it does not meet the criteria until it reaches ```bias_max``` or or it meets the criteria using an if statement. This seems better to me.

**Comment by colemonnahan** [comment]
I really think we need to do some side investigations to bias adjustment. Can we look back at the models from last time and compare the bias adjustment parameters between scenarios? I'm just wondering if it would not be better/simpler/possible to use the same set of values for all scenarios, and that bias adjustment could be done as an initial setup of the simulation and then hard coded into the model files before running.

I'm afraid we're wasting a ton of time/effort on it when we have not really checked what we're gaining. I think investigating the parameters from last time would be a good first step.

**Comment by kellijohnson** [comment]
I don't have a saved version of the results from last round but I do 
have three scenarios for when natural mortality was being estimated when 
it was stationary or had an abrupt change (up and down). Here are the 
bias adjustment results with the rows representing 1) average, 2) 
minimum, and 3) maximum:
1940.538344	1989.843663	2008.833463	2010.35566	0.9051
1937.31961	1988.88263	2008.78053	2010.21702	0.8991
1943.1022	1991.33704	2008.89693	2010.43634	0.9137

I do not think we have time to look into justifying bias adjustment runs 
right now and should maybe just proceed with doing 10 runs per scenario 
like last time to get the papers submitted.

**Comment by colemonnahan** [comment]
Kelli: I was trying to propose we don't do bias adjustment for now. We've never seen any other study do it (besides Carey of course) so we may be justified in ignoring it. As I remember that was one of the major hurdles in getting runs actually finished. I say we just pay attention to them as we run and if they're all the same across data scenarios then we can perhaps ditch it.

**Comment by taylori** [comment]
I would be surprised if the estimated settings are all the same across data scenarios, but if they are similar enough, you could definitely simplify how you implement it by hard-wiring the same values based on a small set of runs that converges. I think that would be better than doing no adjustment at all.

100 or 150 seems like a crazy number. Jim Thorson added a much more robust (but slower) partical swarm optimization procedure to the dumb default in SS_fitbiasramp which you can turn on with the input `altmethod = "psoptim"` to `SS_fitbiasramp`. I know that experimenting with one more thing is not what you all need right now, but this might avoid wasting lots of runs unnecessarily. You could also have a default bias-adjustment settings that kick in whenever you don't get adequate number of converged runs out of a reasonable number.

**Comment by seananderson** [comment]
What's the consensus on this? Do people want me to dig into making it possible for bias adjustment to run until a certain number have converged? That in itself should be relatively easy to do. I would also add a maximum number to try before giving up. Maybe I'll just do it. I'm not working on anything else with ss3sim right now.

I can quickly play with `altmethod = "psoptim"` and add that as an option and possibly make it the default if it's not much slower for our models.

That is unless we have good data from our previous studies showing that the adjustment values didn't vary substantially from scenario to scenario. Bonus points if we can quantify that variation to report in our current papers. Kelli's look fairly similar but maybe other scenarios differed more strongly. @Kotkot or @fhurtado do you have yours? If this is fast and easy to check, then great. Otherwise, let's just let the computers chug away. 'Computer time is cheap, human time is expensive', or something like that. ;)

**Comment by colemonnahan** [comment]
@seananderson: I would for to hold off on this because the reason we would need this functionality is if many of the runs were not converging. The more I think about it, the more I think that we need to solve that problem first.

Say we run 100 iterations and get 10 covariance matrices. We then apply the results from those 10 to all runs of that scenario. But what is different about the 10 runs that converged? Something is different and I'm not really sure what it is, but suspect weirdness with recdevs. So if those 10 runs aren't "normal" in a sense - then how do we justify using them for all other runs? I'm not comfortable doing that unless we really understand the lack of convergence better. So I think investigating that is a higher priority. If we can solve the convergence issues (see my recent email) then fixing this issue would just add unnecessary complexity to the code. That's my two cents.

We should keep in mind what Ian said about testing whether we need bias adjustment: we can look at the EM Report.sso file and see the recruits that SS is actually using and make sure the expected value of those is the R0 in the OM for the early years with no fishing. That may be an easier way to test whether we need bias adjust rather than trying to get 10 covariances for models that aren't converging.

**Comment by kellijohnson** [comment]
Here are some results for cod -- 100 iterations using L31 and A31 casefiles with and without bias adjustment. Values are the relative error in parameter estimates. Looks like the biggest differences are for management quantities and not parameter estimates. This was done with 20 bias adjustment runs. Standard deviation of actual bias correction for 20 runs was 0.06. The standard deviation stabilized at about run 7 or 8, before that it was lower.
```
                              bias.no     bias.yes    diffin_re
SSB_MSY_re               0.0276505608 -0.043083727  0.070734288
TotYield_MSY_re          0.0196400901 -0.050356073  0.069996163
SSB_Unfished_re          0.0273154004 -0.043405344  0.070720745
depletion_re             0.0007653822  0.044876325 -0.044110943
L_at_Amin_Fem_GP_1_re    0.0058871500  0.006303000 -0.000415850
L_at_Amax_Fem_GP_1_re    0.0313046970  0.029717955  0.001586742
VonBert_K_Fem_GP_1_re   -0.0407777000 -0.039052350 -0.001725350
CV_young_Fem_GP_1_re     0.0303964100  0.028481100  0.001915310
CV_old_Fem_GP_1_re      -0.2275610000 -0.221107680 -0.006453320
SR_LN_R0_re             -0.0006843850 -0.004380160  0.003695775
SizeSel_1P_1_Fishery_re  0.0468092323  0.041751004  0.005058228
SizeSel_1P_3_Fishery_re  0.0162705686  0.008759667  0.007510902
SizeSel_2P_1_Survey_re   0.0782894019  0.083898900 -0.005609498
SizeSel_2P_3_Survey_re   0.3613517115  0.393991865 -0.032640154

```

**Comment by colemonnahan** [comment]
Unless you're headers are switched, it looks like bias adjustment is making things worse for the management quantities. The parameters it improves on are negligible. Looking at the "improvement" , defined calculated as (abs(bias.no)-abs(bias.yes)) so that negative values mean bias adjustment made estimates worse (less close to 0) -- see below.

Admittedly I haven't run the new models with low data cases, but my experience with the ridiculously high data case is that the estimates of things we care about are unbiased and bias adjustment is not needed. If we can convince ourselves it's not necessary for our studies we'll save a ton of time.
```
parameter	improvement
SSB_MSY_re	-0.015
TotYield_MSY_re	-0.031
SSB_Unfished_re	-0.016
depletion_re	-0.044
L_at_Amin_Fem_GP_1_re	0.000
L_at_Amax_Fem_GP_1_re	0.002
VonBert_K_Fem_GP_1_re	0.002
CV_young_Fem_GP_1_re	0.002
CV_old_Fem_GP_1_re	0.006
SR_LN_R0_re	-0.004
SizeSel_1P_1_Fishery_re	0.005
SizeSel_1P_3_Fishery_re	0.008
SizeSel_2P_1_Survey_re	-0.006
SizeSel_2P_3_Survey_re	-0.033
```

**Comment by kellijohnson** [comment]
HAHA yes I double checked the headers to make sure that I was not 
reporting things backwards. The management quantities are worse with 
bias adjustment.
What I didn't show was that when you remove data from the EM (say you 
want to do more and more forecasting to see if it matches the OM) the 
bias adjustment becomes more important and the management quantities 
change from being more biased to being less biased when bias adjustment 
is performed.

In conclusion -- with 100 years of data it makes results worse, with 
less data it makes things better.

**Comment by kellijohnson** [comment]
Oh yeah and the median seems to converge much faster and more reliably than the mean. Why do we use mean?

**Comment by seananderson** [comment]
I agree with everything that's been said about the importance of convergence and how nice it would be to skip bias adjustment if we can. That said, it's quite simple to add an argument for the maximum bias adjustment run tries (`max_bias_tries`) to `ss3sim_base`: https://github.com/ss3sim/ss3sim/commit/13c86a3822d45b30d7cf2a06329570bccbde19b8

That's in a separate branch `bias`, but I can move it into `master` later if we want.

**Comment by taylori** [comment]
I support focusing on convergence issues first.
For cod, the median recruitment is exp((-0.4^2)/2) = 92% of the mean.
With hake, it is exp((-1.4^2)/2) = 37% of the mean.

Ignoring that difference won't necessarily impact your estimates of growth,
but it will certainly impact your estimates of spawning biomass.
-Ian

**Comment by juanlvalero** [comment]
Ian and All
At the last growth meeting we've agreed to look further into the source of bias on derived management quantities (SSB_MSY, depletion), even in deterministic, or close to deterministic runs with lots of data and fixing growth parameters at the truth, and ways to deal with it. Following my call with Ian
